backend/dataall/modules/dataset_sharing/aws/kms_client.py

```diff
@@ -1,10 +1,38 @@
 import logging
+import json
 
 from dataall.base.aws.sts import SessionHelper
 from botocore.exceptions import ClientError
 
 log = logging.getLogger(__name__)
 
+DATAALL_BUCKET_KMS_DECRYPT_SID = 'DataAll-Bucket-KMS-Decrypt'
+DATAALL_KMS_PIVOT_ROLE_PERMISSIONS_SID = 'KMSPivotRolePermissions'
+DATAALL_ACCESS_POINT_KMS_DECRYPT_SID = 'DataAll-Access-Point-KMS-Decrypt'
+
+
+def _remove_malformed_principal(policy: str):
+    log.info(f'Malformed Policy: {policy}')
+    kms_policy = json.loads(policy)
+    statements = kms_policy['Statement']
+    for statement in statements:
+        if statement.get('Sid', 'no-sid') in [
+            DATAALL_BUCKET_KMS_DECRYPT_SID,
+            DATAALL_KMS_PIVOT_ROLE_PERMISSIONS_SID,
+            DATAALL_ACCESS_POINT_KMS_DECRYPT_SID,
+        ]:
+            principal_list = statement['Principal']['AWS']
+            if isinstance(principal_list, str):
+                principal_list = [principal_list]
+            new_principal_list = principal_list[:]
+            for p_id in principal_list:
```

Comment:
I don't understand how the malformed principals are removed here. `AROA` principals are the IAM role canonical names, that does not mean that are "bad principals" right?

Reply:
For these policies everywhere in dataall we use Arns to identify the principal. 
But if the principal is deleted, automatically (i think, it's IAM behaviour) those Arns are replaced by RoleIds.  So, if in data-all created S3 (and KMS) policies you encounter RoleId instead of Arn, it 99% percent means, that the role was deleted.

Of course, to be 100% sure, In case of Malformed policy exception, I need to check all principals by their IDs. Unfortunately boto3 has no instrument to list roles by RoleId (only aws cli has). So, in order to make thinfs easier, I made it that way.

Reply:
Initially we used to write principals with roleIds, but we moved to arns as they are easier to read and understand by end-users. That's why I thought you were changing this. Because we used to work with role ids a lot, there are/were methods to get the role ids directly in the code. You can read the role id from a role by calling the IAM boto3 client:

```
    def get_role_id(accountid, name):
        session = SessionHelper.remote_session(accountid=accountid)
        client = session.client('iam')
        try:
            response = client.get_role(RoleName=name)
            return response['Role']['RoleId']
        except ClientError:
            return None
```
I think we should not clean-up all role ids because the majority of customers use "imported" Buckets and KMS keys and we can wipe out permissions set outside of data.all.

Reply:
Another option: store consumption role IDs in DB. What do you think?

Reply:
On the other hand, there can be several deleted roles at once, so it's better to wipe them all by checking principal ID.